### PR TITLE
[GCP] Fix machine image for cluster creation

### LINKED ISSUE #3027: [GCP] The new provisioner dropped the support for machine image
A user reported that the launch of GCP cluster with machine image fails. We need to investigate if the new provisioner accidentally dropped the support of machine image.

### PR DESCRIPTION
**Pull request by Michaelvll**
Fixes #3027

The `bulkInsert` API fails to create VMs using `machineImage` along with disk parameter override, as it does a check for the source of the disk, which should not be applied when `machineImage` is specified (bug of GCP API).
This PR makes it fallback to the regular `insert` API instead to support the creation of instances using machine image.

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
  - [x] `sky launch -c test-machine-image2 --cloud gcp  --disk-size 256 --num-nodes 2 --gpus A100-80GB:8`; correctly failover through multiple regions
  - [x] `sky launch -c test-machine-image5 --cloud gcp --image-id p
rojects/skypilot-375900/global/machineImages/test-machine-image --disk-size 256 --cpus 2 --num-nodes 4`, successfully create the cluster
   - [x] `sky launch -c test-machine-image5 --cloud gcp --image-id projects/skypilot-375900/global/machineImages/test-machine-image --disk-size 256 --num-nodes 2 --gpus A100-80GB:8` successfuly failver through zones/regions.
- [x] All smoke tests: `pytest tests/test_smoke.py --gcp` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/provision/gcp/instance_utils.py [hunk 1]
```diff
@@ -706,29 +687,110 @@ def create_instances(
...
+        return True
+
     @classmethod
     def _create_instances(
         cls,
         names: List[str],
         project_id: str,
         zone: str,
         config: dict,
-        count: int,
```
Comment (cblmemo) [inline sky/provision/gcp/instance_utils.py:722]: Why remove this?

**Review by cblmemo — COMMENTED** [review]
Thanks for the fix! Left several comments 🫡

File: sky/provision/gcp/instance_utils.py [hunk 2]
```diff
@@ -766,26 +828,25 @@ def _handle_http_error(e):
...
-                         f'Reason: {errors}')
-            _log_errors(errors, operation, zone)
-            return errors
+        for operation in operations:
+            errors = operation.get('error', {}).get('errors', [])
+            if errors:
+                logger.debug('create_instances: Failed to create instances. '
+                             f'Reason: {errors}')
+                _log_errors(errors, operations, zone)
+                return errors
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/provision/gcp/instance_utils.py:844]: Our provisioner will handle all the provisioning failure, and terminate the partially started resources: https://github.com/skypilot-org/skypilot/blob/16cdc7ff5dd39073e356e2b3992745b17103d769/sky/provision/provisioner.py#L173-L211

File: sky/provision/gcp/instance_utils.py (hunk 1 above)
Comment (Michaelvll) [reply sky/provision/gcp/instance_utils.py:722]: This is duplicated with `len(names)`. We should be able to leave the `count` out.

**Review by cblmemo — APPROVED** [review]
LGTM. A nit: should we modify the function name here to sth like `_provision`? Though I'm happy with the current implementation since most of the time it won't trigger single inserts. https://github.com/skypilot-org/skypilot/blob/16cdc7ff5dd39073e356e2b3992745b17103d769/sky/provision/provisioner.py#L53

**Comment by Michaelvll** [comment]
> LGTM. A nit: should we modify the function name here to sth like `_provision`? Though I'm happy with the current implementation since most of the time it won't trigger single inserts.
> 
> https://github.com/skypilot-org/skypilot/blob/16cdc7ff5dd39073e356e2b3992745b17103d769/sky/provision/provisioner.py#L53

Thanks for the suggestion @cblmemo! I suppose the `_bulk_provision` is more about provisioning multiple instances in a single function, it might be fine to keep it this way. : )

**Pull request merged**